yjit/src/yjit.rs

```diff
@@ -1,11 +1,135 @@
+//
+// By declaring external functions and using them, we are asserting the symbols
+// we use have definition in one of the object files we pass to the linker. Declaring
+// a function here that has no definition anywhere causes a linking error.
+//
+// There are more things going on during linking and this section makes a lot of
+// simplifications but hopefully this gives a good enough working mental model.
+//
+// ## Difference from example in the Rustonomicon
+//
+// You might be wondering about why this is different from the [FFI example]
+// in the Nomicon, an official book about Unsafe Rust.
+//
+// There is no #[link] attribute because we are not linking against an external
+// library, but rather implicitly asserting that we'll supply a concrete definition
+// for all C functions we call, similar to how pure C projects put functions
+// across different compilation units and link them together.
+//
+// TODO(alan): is the model different enough on Windows that this setup is unworkable?
+//             Seems prudent to at least learn more about Windows binary tooling before
+//             committing to a design.
+//
+// Alan recommends reading the Nomicon cover to cover as he thinks the book is
+// not very long in general and especially for something that can save hours of
+// debugging Undefined Behavior (UB) down the road.
+//
+// UBs can cause Safe Rust to crash, at which point it's hard to tell which
+// usage of `unsafe` in the codebase invokes UB. Providing safe Rust interface
+// wrapping `unsafe` Rust is a good technique, but requires practice and knowledge
+// about what's well defined and what's undefined.
+//
+// For an extremely advanced example of building safe primitives using Unsafe Rust,
+// see the [GhostCell] paper. Some parts of the paper assume less background knowledge
+// than other parts, so there should be learning opportunities in it for all experience
+// levels.
+//
+// ## Binding generation
+//
+// For the moment declarations on the Rust side are hand written. The code is boilerplate
+// and could be generated automatically with a custom tooling that depend on
+// rust-lang/rust-bindgen. The output Rust code could be checked in to version control
+// and verified on CI like `make update-deps`.
+//
+// Upsides for this design:
+//  - the YJIT static lib that links with miniruby and friends will not need bindgen
+//    as a dependency at all. This is an important property so Ruby end users can
+//    build a YJIT enabled Ruby with no internet connection using a release tarball
+//  - Less hand-typed boilerplate
+//  - Helps reduce risk of C definitions and Rust declaration going out of sync since
+//    CI verifies synchronicity
+//
+// Downsides and known unknowns:
+//  - Using rust-bindgen this way seems unusual. We might be depending on parts
+//    that the project is not committed to maintaining
+//  - This setup assumes rust-bindgen gives deterministic output, which can't be taken
+//    for granted
```

Comment:
My worry here is that the set of outputs would all be semantically equivalent, but not bit by bit identical. For example if it used a hash map internally it might output things in different orders. The recommended way to use the library is apparently running it inside the build script: <https://rust-lang.github.io/rust-bindgen/library-usage.html>. Complete determinism for that use case doesn't seem important.

Reply:
We probably should use bindgen in the intended way and run it inside the build script IMHO. More robust and would presumably enable us to export C struct layouts as well.